services/horizon/internal/expingest/main.go

```diff
@@ -403,3 +419,145 @@ func isCancelledError(err error) bool {
+func newParallelSystems(config Config, workerCount uint, systemFactory func(Config) (System, error)) (*ParallelSystems, error) {
+	if workerCount < 1 {
+		return nil, errors.New("workerCount must be > 0")
+	}
+
+	result := ParallelSystems{
+		workerCount:       workerCount,
+		reingestJobQueue:  make(chan ledgerRange),
+		shutdown:          make(chan struct{}),
+		reingestJobResult: make(chan rangeResult),
+	}
+	for i := uint(0); i < workerCount; i++ {
+		s, err := systemFactory(config)
+		if err != nil {
+			result.Shutdown()
+			return nil, errors.Wrap(err, "cannot create new system")
+		}
+		result.wait.Add(1)
+		go result.work(s)
+	}
+	return &result, nil
+}
+
+func (ps *ParallelSystems) work(s System) {
+	defer func() {
+		s.Shutdown()
+		ps.wait.Done()
+	}()
+	for {
+		select {
+		case <-ps.shutdown:
+			return
+		case reingestRange := <-ps.reingestJobQueue:
+			err := s.ReingestRange(reingestRange.from, reingestRange.to, false)
+			select {
+			case <-ps.shutdown:
+				return
+			case ps.reingestJobResult <- rangeResult{err, reingestRange}:
+			}
+		}
+	}
+}
+
+const (
+	historyCheckpointLedgerInterval = 64
+	minBatchSize                    = historyCheckpointLedgerInterval
+)
+
+func calculateParallelLedgerBatchSize(rangeSize uint32, batchSizeSuggestion uint32, workerCount uint) uint32 {
+	batchSize := batchSizeSuggestion
+	if batchSize == 0 || rangeSize/batchSize < uint32(workerCount) {
+		// let's try to make use of all the workers
+		batchSize = rangeSize / uint32(workerCount)
+	}
+	// Use a minimum batch size to make it worth it in terms of overhead
+	if batchSize < minBatchSize {
+		batchSize = minBatchSize
+	}
+
+	// Also, round the batch size to the closest, lower or equal 64 multiple
+	return (batchSize / historyCheckpointLedgerInterval) * historyCheckpointLedgerInterval
+}
+
+func (ps *ParallelSystems) ReingestRange(fromLedger, toLedger uint32, batchSizeSuggestion uint32) error {
+	batchSize := calculateParallelLedgerBatchSize(toLedger-fromLedger, batchSizeSuggestion, ps.workerCount)
+	pendingJobsCount := 0
+	var result error
+	processSubRangeResult := func(subRangeResult rangeResult) {
+		pendingJobsCount--
+		if result == nil && subRangeResult.err != nil {
+			// TODO: give account of what ledgers were correctly reingested?
+			errMsg := fmt.Sprintf("in subrange %d to %d",
+				subRangeResult.requestedRange.from, subRangeResult.requestedRange.to)
+			result = errors.Wrap(subRangeResult.err, errMsg)
+		}
+	}
+
+	for subRangeFrom := fromLedger; subRangeFrom < toLedger && result == nil; {
+		// job queuing
+		subRangeTo := subRangeFrom + batchSize
+		if subRangeTo > toLedger {
+			subRangeTo = toLedger
+		}
+		subRange := ledgerRange{subRangeFrom, subRangeTo}
+
+		select {
+		case <-ps.shutdown:
+			return errors.New("aborted")
+		case subRangeResult := <-ps.reingestJobResult:
+			processSubRangeResult(subRangeResult)
+		case ps.reingestJobQueue <- subRange:
```

Comment:
It was hard to me to understand the code because subrange calculations are happening even when not needed. Maybe we can move this to a one-liner like:
```suggestion
		case ps.reingestJobQueue <- ledgerRange{from: subRangeFrom, to: min(subRangeFrom + batchSize, toLedger)}:
```

Reply:
Well, that doesn't really work because `min()` doesn't exist and you need to use `min(subRangeFrom+batchSize, toLedger)` in the body of the `case` (`subRangeTo`). So, I don't think that a real solution makes it easier to read.

Reply:
Yes, the assumption was to add `min` function but if you are planning to do https://github.com/stellar/go/pull/2724#discussion_r444326321 we can ignore this comment.

Reply:
After the refactoring you suggested below, it's closer to what you suggest.